x/stake/handler.go

```diff
@@ -37,6 +33,46 @@ func NewHandler(k keeper.Keeper) sdk.Handler {
 
 // Called every block, process inflation, update validator set
 func EndBlocker(ctx sdk.Context, k keeper.Keeper) (ValidatorUpdates []abci.Validator) {
+	cdc := types.MsgCdc
```

Comment:
Is there a reason the queue logic is run in the `EndBlocker` instead of the `BeginBlocker`? I guess it doesn't matter much but the latter would seem a bit more accurate since it means state during execution of transactions within the block will always be consistent (any old redelegations / unbonding delegations will have been deleted).

Reply:
Idk, I did the governance one during EndBlock, so just did this one in EndBlock as well.  Don't mind moving it though, cause I don't think it really matters either way.

What's the reason for doing it in BeginBlock instead of EndBlock?  Also, why would the state of transactions in a block not be consistent if it happens in EndBlock?

Reply:
> any old redelegations / unbonding delegations will have been deleted

This is also true if you move if move it to the endblock of the previous block - so I'd say it's still just as arbitrary. 

The general trend which I've seen in our development is to include more state machine logic in EndBlock - and more special block prep logic in BeginBlock. Although the more I think about it, the more I want to just Merge the two

I'm in favour of leaving in endblock - however I also think it's totally arbitrary

Reply:
I don't think it's arbitrary, if we put the logic in `BeginBlock` instead we now have the invariant that all transactions will see a consistent view of the unbonding delegations / redelegations (where expired ones have been deleted), so we don't need to e.g. check timestamps on slashing.

Reply:
I don't see why it changes the consistency?  All the txs in this block will still see it in the queue, and all the txs in the next block, won't.  You can already get rid of the check in slashing.  Basically, it gives one extra block for evidence to be submitted through txs (which no evidence types are submitted in txs rn anyways, so kinda irrelevant).